### ISSUE TITLE
Ability to close boxes.

### ISSUE THREAD

**Issue by Johannestegner** [goblinfactory/konsole#85]
It would be quite nice to be able to remove a box from the render view. Say, you want to use the box as some type of alert, but when the user have confirmed, it "closes" the box.

This would basically be worthwile to add on the IConsole/IWindow interface (or a new IClosable maybe?), and could be either a dispose thing or a Close() method.
I might create a PR for this if it falls within the scope of the library, so let me know!

**Comment by goblinfactory** [comment]
Hi Johannes
sorry for slow reply, I've been on a long non-tech sabbatical, (2 months) mostly without laptop, ... gettting back to home office tomorrow and I'll take a look at your message and reply soon.

To properly do a a windowed console popup, without blocking background threads runs against the design principles behind the library; specifically that the current windows work and function in-place relative div's in html. and wont in parallel with any other thread that writes to the console.

for example e.g. 
your thread... creates a snall 40 x 5 window called winCompress, and inside that window has a progress bar and some text,
that pushes the current console cursor bottom down 6 lines
allowing your main thread (or any thread)to continue to write to the System.Console ...oblivious to the fact there's a window that's being written to by thread Foo.
If you want to indicate a process has completed, then simply call winCompress.Clear(); winCompress.WriteLine("finished compressing images");
Opening a specific window at location x,y  with width and height, doing something and then "closing" that window, would require that you

a) read the buffer behind where you're about to open the window
b) block any threads from writing to the window that is being overlapped, or update the konsole windowing library to detect overlaps during writing and not write to the overlapped area.

This would require double buffers, one for actual writes, and one for display writes.
This is beyond the scope of the current library.

you can already call myWindow.Clear() which would "clear" any opened window with the currently set fore/background colour.

Lastly;

I need to make some significant breaking changes to Konsole in upcoming versions to support scrolling, and keyboard event handling, and dialogs would need to be worked into that change.

i've already started with some of that work, you can see there's a new base "control" that has been introduced in the beta 7 branch, and that has suspendLayout and resumeLayout methods, that are not in current v6.

The purpose of suspendLayout is to allow for me to create a powerful set of themeable controls that work with the new highspeedwriter, allowing me to fix the missing scrolling functionality in OSX/Nix. 
(long story)

basically, v7, v8 are going to be totally new beasts and will have dialog like behaviors, but there's a lot of work for me to do to make that happen. No shortcuts that I can push into V6.
So V6 api and functionality is pretty much frozen, apart from any bugfixes.  

It's possible I've misunderstood what you mean by "close box" but the above are my thoughts. Happy to keep discussing further.

regards,
Alan

**Comment by Johannestegner** [comment]
Hi there!

Thank you for the response, and totally understandable with vacation. It's open source, you shouldn't be available 24/7! :)

When it comes to the issue, I think that I have just not got the `Clear()` command to work as I expected it to. I will keep on trying and if I understand it right, and get it working, I'll close this issue!
Looking forward to V7 release! And will try to jump in and help out when/where I can :)

**Comment by goblinfactory** [comment]
Clear will clear printed contents of the window, but not clear the window itself if it has a border.
In V6 if if you want to totally hide the window frame, like some type of modal that's now closing, you could do that using 2 windows somethink like the following (pseudo code, so please excuse me if this is quite a bit off, but you should get the gist )
```csharp
  var eraser = new Window(...);
  var modal = eraser.Open("I am the Modal");
  ... do some work
  // now hide modal, call eraser, because eraser is the modal's parent.
  eraser.Clear()
  ```
  
  The reason for the extra eraser window is because otherwise the "parent" of that window is the main console itself, and when you Clear() that, it will clear the whole window, and not just the dialog you just opened.
  
 In later Konsole release I will provide a simple blocking and non blocking modals that will make that easier to do, and still be threadsafe.
 
Be aware that the issue with any floating window is that there is no "hidden line" removal logic, i.e. if box 1 is at absolute position 10, 10 and some other thread is writing to the main console, that threads console writes will happily write over the floating window.

That is something i will address in later releases. 

The design philosophy for early konsole releases was to have each thread allocated very unique window areas that never overlap and the floating window is an artifact that is used to deliver that, it's not meant as a key first class citizen.

The philosophy is to quickly break a screen into a few windows, and then have them run normal C# code that writes to a window that is oblivious, to avoid having to write seperate  server/client console apps when testing libraries and launching them manually or automated, ...this way you can spin up multiple services, and test lots of complex code with a single console application, and see the output cleanly.

Since then Konsole has been used for a lot of applications, games etc.
later releases are aimed at making Konsole a complete cross platform rapid prototyping library. 

cheers, 
Alan

**Comment by Johannestegner** [comment]
Awesome! That will most likely do exactly what I need. Luckily, all of this type of windows that I will need in this application are during "pause" in the rest of the application (i.e., after validation but before run etc), so I will have no issues with other threads overwriting.

All makes sense and I think what you wrote will be more than enough to get me going!

Thank you very much for the help :)

**Comment by goblinfactory** [comment]
Ah! great!...glad to be able to help...

Are you using the latest alpha v7 release, it's actually quite stable?
There are a lot of overloads for window creation and colouring (theming)
I've not written documentation yet, because It's a really massive change.
A

**Comment by Johannestegner** [comment]
@goblinfactory currently on the v6 branch, but I did take a look at the new features in v7 and it looks very promising! :)
I'll try upgrade the current project to v7 and see how it fares! :)

**Comment by goblinfactory** [comment]
Is your project open so that I could look at how you're using Konsole? That would be very helpful for me in determining which features to work on, and also what examples and documentation to write to help users of the library use it idiomatically (as designed) and get the most from the library :D

**Comment by Johannestegner** [comment]
Sadly not this one. But since finding Konsole, I have been considered using it for a few other projects that I have, which are more open. Will gladly let you know when I got something to show :)

I do love using and building CLI applications, so this will likely give me much fun during the evenings ;D

**Comment by goblinfactory** [comment]
ah! Understood. Great to hear you want to use it in your own projects and other new  ideas.
Cheers, 
Alan
